lib/utils/BaseReporter.js

```diff
@@ -63,45 +66,76 @@ class BaseReporter extends events.EventEmitter {
         })
 
         this.on('runner:start', (runner) => {
-            if (!this.stats.runner[runner.cid]) {
-                this.stats.runner[runner.cid] = {
-                    start: new Date(),
-                    capabilities: runner.capabilities,
-                    config: runner.config,
-                    tests: []
-                }
-            }
+            this.getSpecStats(runner)
```

Comment:
So, my thinking is that reporters might want either to handle each event as it comes in (in which case, we just forward the payload from the BaseReporter to each reporter), but that some reporters might alternatively only want a summary of everything at the end. The Junit reporter needs to do nothing until the end (it can't, as the XML requires a summary of the test passes & failures, etc). If it collects the data itself based on the events, that's fine until another reporter starts to do exactly the same thing. It seems like a responsibility of the BaseReporter to be able to summarise everything. 

If it's to do this, the BaseReporter needs to be able to associate every event payload with the current spec, suite or test. For example, when `runner:command` is fired, the payload includes enough data (`title`, `parent`, `cid`, `specHash`) to identify where in the large `this.stats` object to store the payload.

The reason `getTestStats()` (and others) are horrible methods is that they have two purposes - in the case of `test:pass` it will return the current test, and we can set `state = 'pass'`, but in the case of `test:start` all we need to do is to create the test record in the right place within `this.stats`, and set `start: new Date()`. Of course, I could tweak it slightly so that `getTestStats()` only returns an object if it exists, then each caller would have to do something like:

```js
        this.on('test:start', (test) => {
            const testStats = this.getTestStats(test) || this.createTestStats(test)
            testStats.state = 'pass'
            this.stats.passes++
            this.propagateEvent('test:pass', test)
        })
```

I suppose memory is also a concern here - keeping so much test output might not work for long-running tests.

I don't like the design I've come up with - I agree with you comments. But perhaps you understand the use-case better now? Let me know your thoughts.